server/src/main/java/io/deephaven/server/session/SessionServiceGrpcImpl.java

```diff
@@ -290,24 +316,138 @@ public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(final ServerCall<Re
+                            } else {
+                                log.error().append(cause).endl();
+                            }
+
+                            call.close(sre.getStatus(), sre.getTrailers());
+                            return new ServerCall.Listener<>() {};
+                        } else {
+                            log.error().append("Failed to invoke auth method: ").append(e).endl();
+                            return onFailure.get();
+                        }
+                    }
+                }
+
+                final ServerCall.Listener<ReqT> delegate = serverCallHandler.startCall(innerCall, headers);
+                return new ForwardingServerCallListener.SimpleForwardingServerCallListener<>(delegate) {
+                    @Override
+                    public void onMessage(ReqT message) {
+                        // authorize requests now if the client was required to send a single request
+                        if (!isClientStreaming && authMethod != null) {
+                            try {
+                                authMethod.invoke(authWiring, authContext, message);
+                            } catch (IllegalAccessException | InvocationTargetException e) {
+                                final Throwable cause = e.getCause();
+                                if (cause instanceof StatusRuntimeException) {
+                                    final StatusRuntimeException sre = (StatusRuntimeException) cause;
+                                    Status.Code status = sre.getStatus().getCode();
+                                    if (status.equals(Status.Code.UNAUTHENTICATED)) {
+                                        log.info().append(serviceName).append(".").append(methodName)
+                                                .append(": request unauthenticated").endl();
+                                    } else if (status.equals(Status.Code.PERMISSION_DENIED)) {
+                                        log.info().append(serviceName).append(".").append(methodName)
+                                                .append(": permission denied ").endl();
+                                    } else {
+                                        log.error().append(cause).endl();
+                                    }
+
+                                    call.close(sre.getStatus(), sre.getTrailers());
+                                } else {
+                                    log.error().append("Failed to invoke auth method: ").append(e).endl();
+                                    onFailure.get();
+                                }
+
+                                return;
+                            }
+                        }
+
+                        // if we're still here, then it is safe to continue with the request
+                        delegate().onMessage(message);
+                    }
+                };
+            };
```

Comment:
There's a lot of logic here - I wonder if it makes sense to break these up into classes instead of anonymous.

Reply:
I am still worried about the potential fragility here. I think we want a good way to unit test this logic (as opposed to integration tests), and I don't see a good way to do that as-is.

It would be great to be able to test the reflection in isolation.

Reply:
I've replaced this with service method overrides instead of injection.